File: quickSort/quickSort.cpp
// Quick Sort in C++
// using Median of Three method to find suitable pivot for partitioning

#include <iostream>
#include <vector>
using namespace std;

template < typename T,
            typename Comp = less<T> >
T median_of_three(T first, T second, T third, Comp mLess = Comp()){
    vector<T> temp = {first, second, third};

    // insertion sort in this function
    for(int step = 1; step < (int)temp.size(); ++step){
        T value = temp[step];
        int j = step -1;
        while(j >= 0 && mLess(value, temp[j])){
            temp[j +1] = temp[j];
            --j;
        }
        temp[j +1] = value;
    }
    // end insertion sort

    return temp[1];
}

template <typename T,
            typename Comp = less<T> >
int partition(vector<T> &arr, int start, int stop, Comp mLess = Comp()){
    int mid = (start + stop) >> 2;
    T pivot = median_of_three(arr[start], arr[mid], arr[stop]);

    int left = --start;
    int right = ++stop;

    while(left < right){

        do{
            ++left;
        }
        while(mLess(arr[left], pivot));

        do{
            --right;
        }
        while(mLess(pivot, arr[right]));

        if(left < right) swap(arr[left], arr[right]);
        else             return right;
    }
    return right;
}

template <typename T,
            typename Comp = less<T> >
void quickSortAux(vector<T> &arr, int start, int stop, Comp mLess = Comp()){
    if(start < stop){
        int pivot = partition(arr, start, stop, mLess);
        quickSortAux(arr, start, pivot, mLess);
        quickSortAux(arr, pivot +1, stop, mLess);
    }
}

template <typename T,
            typename Comp = less<T> >
void quickSort(vector<T> &arr, Comp mLess = Comp()){
    quickSortAux(arr, 0, (int)arr.size() -1, mLess);
}

File: quickSort/quickSort.py
# Quick Sort in python
from statistics import median

def quickSort(array, mLess = lambda x, y : x < y):
    quickSortAux(array, 0, len(array) -1, mLess)
    return

def quickSortAux(array, start, stop, mLess = lambda x, y : x < y):
    if start < stop:
        pivot = partition(array, start, stop, mLess)
        quickSortAux(array, start, pivot, mLess)
        quickSortAux(array, pivot +1, stop, mLess)
    return

def partition(array, start, stop, mLess = lambda x, y : x < y):
    mid = (start + stop) //2
    pivot = median([array[start], array[mid], array[stop -1]])
    
    left = start -1
    right = stop +1

    while(left < right):
        while 1:
            left += 1
            if(not mLess(array[left], pivot)):
                break
        while 1:
            right -= 1
            if(mLess(array[right], pivot) or (array[right] == pivot)):
                break
        if(left < right):
            array[left], array[right] = array[right], array[left]
        else:
            return right
    return right